scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+            try: () => response.json(),
+            catch: () => new JSONError(),
+          })
+        ),
+        Effect.flatMap(parsePokemon)
+      ),
+  }),
+  Effect.runPromise
+);
+```
+
+notes:
+now we can provide the pokemon client implementation before passing the result to runPromise. Here our implementation of the getById methos is identical to the logic of our getPokemon function previously, but now we could change the data source or adding caching and our downstream usage of the client doesn't need to change at all.
+
+---
+
+```ts
+const getPokemon = (id: number) =>
+  Effect.gen(function* (_) {
+    const client = yield* _(PokemonClient);
+    return yield* _(client.getById(id));
+  }).pipe(
+    Effect.catchAll(() => Effect.succeed({ name: "default", weight: 0 }))
+  );
+```
+
+notes:
+With generators we get the client by yielding the tag and return the yielded result of getById, then provide the service the same as we did just now in the pipe version.
+
+---
+
+## Logging
+
+notes:
+Our final topic will be logging. Effect provides a powerful set of logging tools to use observe your program.
+
+---
+
+```ts
+// type: Effect<never, never, void>
+const program = Effect.log("Application started")
+ 
+Effect.runSync(program)
+/*
+Output:
+timestamp=2023-07-05T09:14:53.275Z level=INFO fiber=#0 message="Application started"
+*/
+```
+
+notes:
+the log function logs the given message to the console along with some additional information and returns a void effect
+
+---
+
+```ts
+const logLevels = pipe(
+  Effect.log("info by default"),
+  Effect.flatMap(() => Effect.logDebug("debug")),
+  Effect.flatMap(() => Effect.logInfo("info")),
+  Effect.flatMap(() => Effect.logWarning("warning")),
+  Effect.flatMap(() => Effect.logError("error")),
+  Effect.flatMap(() => Effect.logFatal("fatal"))
+);
+
+Effect.runSync(logLevels);
+```
+```
+timestamp=2023-07-25T04:30:54.447Z level=INFO fiber=#1 message="info by default"
+timestamp=2023-07-25T04:30:54.447Z level=INFO fiber=#1 message=info
+timestamp=2023-07-25T04:30:54.447Z level=WARN fiber=#1 message=warning
+timestamp=2023-07-25T04:30:54.448Z level=ERROR fiber=#1 message=error
+timestamp=2023-07-25T04:30:54.448Z level=FATAL fiber=#1 message=fatal
+```
+
+notes:
+Effect has different log levels that can be used to categorize your logs. You may notice that the logDebug call was not printed to the console, and this is intentional. By default all debug logs are not printed.
+
+---
+
+```ts
+import { Effect, Logger, LoggerLevel, pipe } from "effect";
+
+const logDebug = pipe(
+  Effect.logDebug("debug"),
+  Logger.withMinimumLogLevel(LoggerLevel.Debug)
+);
+
+Effect.runSync(logDebug);
+```
+```
+timestamp=2023-07-25T04:33:07.743Z level=DEBUG fiber=#2 message=debug
+```
+
+notes:
+Using the withMinimumLogLevel function from the Logger namespace we can manually adjust this setting. In this case we can lower it to now show debug logs, but if you only set a higher minimum level you can as well.
```

Comment:
`namespace` => `module` otherwise the uninformed listener will think we use ts namespaces